### PR TITLE
[PyTorch Edge] share tensors in mobile with new api

### PR DESCRIPTION
**Pull request by cccclai**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57888 [PyTorch Edge] bytecode version bump to v5 and enable share constant table
* **#58182 [PyTorch Edge] share tensors in mobile with new api**

As title, the v5 model format will be
```
(base) chenlai@chenlai-mp reuse_constant % zipinfo /Users/chenlai/Documents/pytorch/reuse_constant/tmp/zip/script_module_v5_unify.ptl
Archive:  /Users/chenlai/Documents/pytorch/reuse_constant/tmp/zip/script_module_v5_unify.ptl
Zip file size: 3120 bytes, number of entries: 7
-rw----     0.0 fat       77 bl stor 80-000-00 00:00 script_module_v4_unify/data.pkl
-rw----     0.0 fat      240 bl defN 80-000-00 00:00 script_module_v4_unify/code/__torch__/___torch_mangle_5.py
-rw----     0.0 fat      422 bl defN 80-000-00 00:00 script_module_v4_unify/code/__torch__/___torch_mangle_5.py.debug_pkl
-rw----     0.0 fat       64 bl stor 80-000-00 00:00 script_module_v4_unify/constants/140245072983168.storage
-rw----     0.0 fat      172 bl stor 80-000-00 00:00 script_module_v4_unify/constants.pkl
-rw----     0.0 fat      678 bl stor 80-000-00 00:00 script_module_v4_unify/bytecode.pkl
-rw----     0.0 fat        2 bl stor 80-000-00 00:00 script_module_v4_unify/version
7 files, 1655 bytes uncompressed, 1453 bytes compressed:  12.2%
```
bytecode.pkl is:
```
(5,
 ('__torch__.___torch_mangle_5.TestModule.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('LOADC', 0, 0),
     ('LOADC', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('LOADC', 1, 0),
     ('OP', 1, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::add', 'int'), ('aten::add', 'Scalar'))),
   ('constants',
    (torch._utils._rebuild_tensor_v2(pers.obj(('storage',
          torch.DoubleStorage,
          '140245072983168.storage',
          'cpu',
          8),),
       0,
       (2, 4),
       (4, 1),
       False,
       collections.OrderedDict()),
     1)),
   ('types', ()),
   ('register_size', 2)),
  (('arguments',
    ((('name', 'self'),
      ('type', '__torch__.___torch_mangle_5.TestModule'),
      ('default_value', None)),
     (('name', 'y'), ('type', 'int'), ('default_value', None)))),
   ('returns',
    ((('name', ''), ('type', 'Tensor'), ('default_value', None)),)))))
```

constants.pkl is:
```
(torch._utils._rebuild_tensor_v2(pers.obj(('storage', torch.DoubleStorage, '140245072983168.storage', 'cpu', 8),),
   0,
   (2, 4),
   (4, 1),
   False,
   collections.OrderedDict()),)
```

Both tensors will refer to the tensor in at the path `script_module_v4_unify/constants/140245072983168.storage`.

## Note
According to unify format, all tensors will be written to the folder `.data`, however, torch.jit.load() can't handle the unified format at this moment, so this change will write tensors at the `constants` folders, and mobile will write/read tensors from `constants` folder. such that the model can be interpreted by both jit and mobile.

Differential Revision: [D28375257](https://our.internmc.facebook.com/intern/diff/D28375257/)

### CODE REVIEW

File: torch/csrc/jit/mobile/backport_manager.cpp
Comment (raziel) [inline torch/csrc/jit/mobile/backport_manager.cpp:51]: is substring safe here?
otherwise, should be be strictly checking for the full path (and thus use the set's find function instead)?

File: torch/csrc/jit/mobile/import.h
Comment (raziel) [inline torch/csrc/jit/mobile/import.h:92]: maybe IsTensorIn...

some docs would be nice too.

**Review by raziel — COMMENTED** [review]
Thanks!

I suggest focusing the description on the goal of the changes rather than on whether this is the new or old api.

I'm not familiar with the changes in serialization so @Lilyjjo and @suo should take a closer look (plus @iseeyuan for the mobile specific details).

File: caffe2/serialize/versions.h
Comment (iseeyuan) [inline caffe2/serialize/versions.h:70]: Are we bumping the version in the same PR?

File: torch/csrc/jit/serialization/export_module.cpp
Comment (iseeyuan) [inline torch/csrc/jit/serialization/export_module.cpp:355]: Add comment on this arg, like the args above.

File: torch/csrc/jit/serialization/pickler.cpp
Comment (iseeyuan) [inline torch/csrc/jit/serialization/pickler.cpp:314]: Looks a good change to me. @suo and @Lilyjjo please confirm if it's good.

**Review by iseeyuan — COMMENTED** [review]
I think it's good to share the API in general. Would the tests be on top of this PR?

File: torch/csrc/jit/serialization/pickler.cpp
Comment (cccclai) [reply torch/csrc/jit/serialization/pickler.cpp:314]: It's a temporary fix and @Lilyjjo addressed it in the stack

File: torch/csrc/jit/mobile/backport_manager.cpp
Comment (cccclai) [reply torch/csrc/jit/mobile/backport_manager.cpp:39]: Added in the latest commit.

File: torch/csrc/jit/serialization/export_module.cpp
Comment (Lilyjjo) [inline torch/csrc/jit/serialization/export_module.cpp:352]: Just a note, this change will introduce jitter between saves & loads of the same model since now the tensor names are no longer always the same 0/1/2/3/4 and instead are based off of the tensor's storage's c-memory pointer

Comment (Lilyjjo) [reply torch/csrc/jit/serialization/export_module.cpp:352]: @cccclai is there a reason why you didn't also set the `tensor_cdata_naming_scheme` to true for the writeArchive call on `:334`? I might be misunderstanding the flow of mobile serialization, but I know that tensors can be written out to a `data/` folder instead of the `constants/` folder and you'd want those also to use the `tensor_cdata_naming_scheme` and `constants/` location for those tensors also

Comment (suo) [reply torch/csrc/jit/serialization/export_module.cpp:352]: Ah this is a format thing. We actually write tensors to two locations today: the data folder, which contains tensors that are explicitly part of the module state (like parameters and buffers), and the constants folder, which contains tensors that are `prim::Constants` in the graph. 

As far as I understand the major goal of this work is to share the constants between the bytecode and TS code, since the data folder is already shared. Is this right @cccclai?

Comment (cccclai) [reply torch/csrc/jit/serialization/export_module.cpp:352]: Good question! At this moment, the majority (probably all, I'm just being conservative) of duplicated tensors are between `constants` and `bytecode`. Also, for mobile, there is an `optimize_for_mobile()` step to convert tensors from parameters (in `data`) to `constants`, and ideally, there shouldn't be any duplicated tensors between `data` and `bytecode` after `optimized_for_mobile()` step. Given that it's unlikely to have duplication, I think we can wait until the unify format

**Review by suo — COMMENTED** [review]
I only looked at the `export_module.cpp` changes but they seem small and self-contained, lgtm. I think the smallness of the changes is a good sign we are on the right track with regard to unification!

Will leave it to @raziel or @iseeyuan to accept

File: torch/csrc/jit/mobile/backport_manager.cpp
Comment (iseeyuan) [reply torch/csrc/jit/mobile/backport_manager.cpp:55]: Good point. A record includes the path but expluded_file does not. Is it possible that there are two files with the same name but in different paths?

File: torch/csrc/jit/serialization/export_module.cpp
Comment (cccclai) [reply torch/csrc/jit/serialization/export_module.cpp:352]: @suo yeah this statement is correct.
> the major goal of this work is to share the constants between the bytecode and TS code, since the data folder is already shared

File: torch/csrc/jit/serialization/pickler.cpp
Comment (cccclai) [reply torch/csrc/jit/serialization/pickler.cpp]: Rebase and there is no change in `pickler.cpp` now.

File: torch/csrc/jit/serialization/export_module.cpp
Comment (cccclai) [reply torch/csrc/jit/serialization/export_module.cpp]: Add it in the latest commit.

File: caffe2/serialize/versions.h
Comment (iseeyuan) [reply caffe2/serialize/versions.h:70]: Chatted offline. It's for testing purpose. Please remember to change it back to 4 before landing it.

Comment (cccclai) [reply caffe2/serialize/versions.h:70]: Ideally yeah, consider the change is quite minimum and all the test in place.

File: torch/csrc/jit/serialization/export_module.cpp
Comment (iseeyuan) [reply torch/csrc/jit/serialization/export_module.cpp:352]: @suo Yes, we shared data (parameters) but not constants between jit and bytecode. For this PR we also share constants.

File: torch/csrc/jit/mobile/backport_manager.cpp
Comment (cccclai) [reply torch/csrc/jit/mobile/backport_manager.cpp:55]: Yeah it's a good point. Given our current model schema, it should be fine. But I agree full path will be safer. I add a bit more context here and more comment in the source code and hope it's easier to understand. `reader.getAllRecords()` will return a file list, say:
```
script_module_v4_unify/data.pkl
script_module_v4_unify/code/__torch__/___torch_mangle_5.py
script_module_v4_unify/code/__torch__/___torch_mangle_5.py.debug_pkl
script_module_v4_unify/constants/0
script_module_v4_unify/constants.pkl
script_module_v4_unify/bytecode/0
script_module_v4_unify/bytecode.pkl
script_module_v4_unify/version
```
If we want to skip writing bytecode archive, we need to skip both `script_module_v4_unify/bytecode/0` (the actual tensor data) and `script_module_v4_unify/bytecode.pkl`. We currently don't have a way to know how many tensors in total without reading `bytecode.pkl` + `bytecode` folder together. I'm looking at an easier way to load pickle file, since it's just a zip folder and it shouldn't need any pytorch related knowledge. After it's done, we can leverage it here. It can be part of the work in T90180710. Let me add it in the comment so we can keep track of it.

File: caffe2/serialize/versions.h
Comment (cccclai) [reply caffe2/serialize/versions.h]: Move the version bump to #57888

File: torch/csrc/jit/mobile/import.h
Comment (cccclai) [reply torch/csrc/jit/mobile/import.h]: Rename to `isTensorInBytecodeArchive` in latest version. As a note, it is *not* `isTensorInConstantArchive`. I realize that there are always tensors in `constants` archive, but not `bytecode` archive in the future model, so I change the logic.

File: torch/csrc/jit/mobile/backport_manager.cpp
Comment (cccclai) [reply torch/csrc/jit/mobile/backport_manager.cpp]: Use regex to match instead in the latest commit.

File: torch/csrc/jit/mobile/import.h
Comment (cccclai) [reply torch/csrc/jit/mobile/import.h]: In terms of why unit test didn't catch it? Likely all tensors in existing testing models from `bytecode` are exactly the same as `constants`...anyway, it's better to use `isTensorInBytecodeArchive` logic.

File: torch/csrc/jit/mobile/backport_manager.cpp
Comment (cccclai) [reply torch/csrc/jit/mobile/backport_manager.cpp]: Ok internal CI doesn't like regex. Switch back with more strict substring searching

File: torch/csrc/jit/serialization/export_module.cpp
Comment (Lilyjjo) [reply torch/csrc/jit/serialization/export_module.cpp]: Awesome, this makes sense then! :) 🚢 🚢

File: torch/csrc/jit/mobile/backport_manager.cpp
Comment (iseeyuan) [inline torch/csrc/jit/mobile/backport_manager.cpp:65]: Nice.

**Review by iseeyuan — APPROVED** [review]
LGTM. Thanks!

**Review by raziel — APPROVED** [review]
Thanks a lot @cccclai

**Pull request closed without merging**
